connstr/connstr.go

```diff
@@ -0,0 +1,683 @@
+func (P ConnectionParams) StringNoClass() string {
+	return P.string(false, false)
+}
+
+// StringWithPassword returns the string representation of ConnectionParams (as String() does),
+// but does NOT obfuscate the password, just prints it as is.
+func (P ConnectionParams) StringWithPassword() string {
+	return P.string(true, true)
+}
+
+func (P ConnectionParams) string(class, withPassword bool) string {
+	q := NewParamsArray(32)
+	s := P.ConnClass
+	if !class {
+		s = ""
+	}
+	q.Add("connectionClass", s)
+
+	q.Add("user", P.Username)
+	if withPassword {
+		q.Add("password", P.Password.Secret())
+		q.Add("newPassword", P.NewPassword.Secret())
+	} else {
+		q.Add("password", P.Password.String())
+		if !P.NewPassword.IsZero() {
+			q.Add("newPassword", P.NewPassword.String())
+		}
+	}
+	s = "local"
+	tz := P.Timezone
+	if tz != nil && tz != time.Local {
+		s = tz.String()
+	}
+	q.Add("timezone", s)
+	B := func(b bool) string {
+		if b {
+			return "1"
+		}
+		return "0"
+	}
+	q.Add("poolMinSessions", strconv.Itoa(P.MinSessions))
+	q.Add("poolMaxSessions", strconv.Itoa(P.MaxSessions))
+	q.Add("poolIncrement", strconv.Itoa(P.SessionIncrement))
+	q.Add("sysdba", B(P.IsSysDBA))
+	q.Add("sysoper", B(P.IsSysOper))
+	q.Add("sysasm", B(P.IsSysASM))
+	q.Add("standaloneConnection", B(P.StandaloneConnection))
+	q.Add("enableEvents", B(P.EnableEvents))
+	q.Add("heterogeneousPool", B(P.Heterogeneous))
+	q.Add("prelim", B(P.IsPrelim))
+	q.Add("poolWaitTimeout", P.WaitTimeout.String())
+	q.Add("poolSessionMaxLifetime", P.MaxLifeTime.String())
+	q.Add("poolSessionTimeout", P.SessionTimeout.String())
+	q.Values["onInit"] = P.onInitStmts
+	q.Add("configDir", P.ConfigDir)
+	q.Add("libDir", P.LibDir)
+	//return quoteRunes(P.Username, "/@") + "/" + quoteRunes(password, "@") + "@" + P.CommonParams.ConnectString + "\n" + q.String()
+
+	var buf strings.Builder
+	buf.WriteString(P.CommonParams.ConnectString)
+	buf.WriteByte('\n')
+	q.WriteTo(&buf)
+	return buf.String()
+}
+
+// Parse parses the given connection string into a struct.
+func Parse(dataSourceName string) (ConnectionParams, error) {
+	P := ConnectionParams{
+		StandaloneConnection: DefaultStandaloneConnection,
+		CommonParams: CommonParams{
+			Timezone: time.Local,
+		},
+		ConnParams: ConnParams{
+			ConnClass: DefaultConnectionClass,
+		},
+		PoolParams: PoolParams{
+			MinSessions:      DefaultPoolMinSessions,
+			MaxSessions:      DefaultPoolMaxSessions,
+			SessionIncrement: DefaultPoolIncrement,
+			MaxLifeTime:      DefaultMaxLifeTime,
+			WaitTimeout:      DefaultWaitTimeout,
+			SessionTimeout:   DefaultSessionTimeout,
+		},
+	}
+
+	var paramsString string
+	dataSourceName = strings.TrimSpace(dataSourceName)
+	if i := strings.IndexByte(dataSourceName, '\n'); i >= 0 {
+		dataSourceName, paramsString = strings.TrimSpace(dataSourceName[:i]), strings.TrimSpace(dataSourceName[i+1:])
+	}
+	var q url.Values
+
+	if !strings.HasPrefix(dataSourceName, "oracle://") {
+		// Not URL
+		q = make(url.Values, 32)
```

Comment:
The 32 here is used above as well and should probably be a constant of some kind?